src/Symfony/Component/Messenger/Middleware/HandleMessageMiddleware.php

```diff
@@ -60,7 +62,10 @@ public function handle(Envelope $envelope, StackInterface $stack): Envelope
 
             try {
                 $handler = $handlerDescriptor->getHandler();
-                $handledStamp = HandledStamp::fromDescriptor($handlerDescriptor, $handler($message));
+                if ($handler instanceof ConfigureAutoAckInterface && $handler->disableAutoAck($envelope)) {
+                    $envelope = $envelope->with(new DelayedAckStamp());
+                }
```

Comment:
Maybe this is not needed? The handler (userland) could add the `DelayedAckStamp` instead of implementing the `ConfigureAutoAckInterface`.
In short: User adds a `DelayedAckStamp` stamp, to the envelope, in that case the worker don't ack.,

In your example:

```diff
-final class HelloMessageHandler implements MessageHandlerInterface, ConfigureAutoAckInterface, EventSubscriberInterface
+final class HelloMessageHandler implements MessageHandlerInterface, EventSubscriberInterface
 {
     public function __invoke(HelloMessage $message, Envelope $envelope)
     {
-        if ($this->disableAutoAck($envelope)) {
+        if ($envelope->last(ReceivedStamp::class)->getTransportName() === 'async') {
+           $envelope = $envelope->with(new DelayedAckStamp());
            $this->logger->info('Add message to buffer.');
            $this->buffer[] = $envelope;
            $this->flushIfNeeded();
 
             return;
         }
 
         $this->logger->info('process message.');

         // Do regular processing
     }

-    public function disableAutoAck(Envelope $envelope): bool
-    {
-        // This handler could be used by many transports.
-        // But only the async one should be manually controlled // bufferised 
-        return $envelope->last(ReceivedStamp::class)->getTransportName() === 'async';
-    }
 ...
 }

Reply:
This does not work because of immutability 😢